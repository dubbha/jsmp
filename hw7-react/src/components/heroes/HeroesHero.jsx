import React from 'react';
import PropTypes from 'prop-types';

const HeroesHero = ({ hero, setSelectedHero, className }) => (
  <li onClick={setSelectedHero} className={className}>
    <span className="badge">{hero.id}</span> {hero.name}
  </li>
);

HeroesHero.propTypes = {
  hero: PropTypes.object.isRequired,
  setSelectedHero: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired,
};

export default HeroesHero;
