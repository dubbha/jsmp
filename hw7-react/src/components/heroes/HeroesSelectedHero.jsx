import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const HeroesSelectedHero = ({ hero }) => (
  <div className="selectedHero">
    <h2 className="selectedHero__header">
      <span className="uppercase">{hero.name}</span> is my hero
    </h2>
    <Link to={`/hero-details/${hero.id}`} className="selectedHero__link">View Details</Link>
  </div>
);

HeroesSelectedHero.propTypes = {
  hero: PropTypes.object.isRequired,
};

export default HeroesSelectedHero;
