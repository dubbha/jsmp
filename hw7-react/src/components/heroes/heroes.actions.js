export const actionTypes = {
  HERO_UPDATE: 'HERO_UPDATE',
  HERO_ADD: 'HERO_ADD',
  HERO_DELETE: 'HERO_DELETE',
};

export const updateHero = hero => ({
  type: actionTypes.HERO_UPDATE,
  hero,
});

export const addHero = hero => ({
  type: actionTypes.HERO_ADD,
  hero,
});

export const deleteHero = id => ({
  type: actionTypes.HERO_DELETE,
  id,
});
