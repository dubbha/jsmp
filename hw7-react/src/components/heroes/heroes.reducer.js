export const initialState = [
    { id: 11, name: 'React' },
    { id: 12, name: 'Component' },
    { id: 13, name: 'State' },
    { id: 14, name: 'Router' },
    { id: 15, name: 'Redux' },
    { id: 16, name: 'Action' },
    { id: 17, name: 'Reducer' },
    { id: 18, name: 'Store' },
    { id: 19, name: 'Connect' },
    { id: 20, name: 'Provider' },
];

const heroesReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'HERO_ADD':
      return [...state, { ...action.hero }];
    case 'HERO_UPDATE':
      return state.map(el => (el.id === action.hero.id ? { ...action.hero } : el));
    case 'HERO_DELETE':
      return state.filter(el => el.id !== action.id);
    default:
      return state;
  }
};

export default heroesReducer;
