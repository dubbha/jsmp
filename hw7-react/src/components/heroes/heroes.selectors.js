import { createSelector } from 'reselect';

export const heroesSelector = state => state.heroes;

export const heroIdSelector = (state, props) => +props.match.params.id;

export const heroSelector = (state, props) => (
  state.heroes.filter(hero => hero.id === +props.match.params.id)[0]
);

export const heroReselector = createSelector(
  [heroesSelector, heroIdSelector],
  (heroes, id) => (
    heroes.filter(hero => hero.id === id)[0]
  ),
);