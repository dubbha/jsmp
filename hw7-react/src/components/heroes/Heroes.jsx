import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import HeroesHero from './HeroesHero';
import HeroesSelectedHero from './HeroesSelectedHero';
import * as heroesSelectors from './heroes.selectors';
import './style.sass';

export class Heroes extends React.Component {
  static propTypes = {
    heroes: PropTypes.array.isRequired,
  };
  constructor(props) {
    super(props);

    this.state = {
      title: 'Heroes',
      selectedHero: null,
    };
  }

  setSelectedHero = (hero) => {
    this.setState({
      selectedHero: hero
    });
  }

  render() {
    const { heroes } = this.props;
    const { selectedHero } = this.state;
    return (
      <div className="heroes">
        <h2>My Heroes</h2>
        <ul className="heroes">
          {heroes.map(hero => (
              <HeroesHero
                hero={hero}
                key={hero.id}
                setSelectedHero={() => this.setSelectedHero(hero)}
                className={(this.state.selectedHero && this.state.selectedHero.id === hero.id)
                              ? 'selected' : '' }
              />))}
        </ul>
        { selectedHero ? (<HeroesSelectedHero hero={selectedHero} />) : '' }
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  heroes: heroesSelectors.heroesSelector(state),
});

export default connect(
  mapStateToProps,
)(Heroes);
