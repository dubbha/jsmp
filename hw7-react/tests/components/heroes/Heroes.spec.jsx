import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Heroes, mapStateToProps } from 'components/heroes/Heroes';

jest.mock('components/heroes/heroes.selectors', () => ({
  heroesSelector: state => state.heroes,
}));

describe('Heroes', () => {
  it('should render heroes list', () => {
    const props = {
      heroes: [
        { id: 3, name: 'Optimus Drie' },
        { id: 4, name: 'Optimus Fier' },
        { id: 5, name: 'Optimus Funf' },
      ],
    };
    const wrapper = shallow(<Heroes {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should map state to props', () => {
    const state = {
      heroes: [
        { id: 3, name: 'Optimus Drie' },
        { id: 4, name: 'Optimus Fier' },
        { id: 5, name: 'Optimus Funf' },
      ]
    };

    const props = { heroes: [] };

    expect(mapStateToProps(state, props)).toEqual({
      heroes: [
        { id: 3, name: 'Optimus Drie' },
        { id: 4, name: 'Optimus Fier' },
        { id: 5, name: 'Optimus Funf' },
      ]
    });
  });

  it('should select a hero on a child click', () => {
    const props = {
      heroes: [
        { id: 3, name: 'Optimus Drie' },
        { id: 4, name: 'Optimus Fier' },
        { id: 5, name: 'Optimus Funf' },
      ],
    };
    const wrapper = shallow(<Heroes {...props} />);
    wrapper.find('HeroesHero').at(1).prop('setSelectedHero')();
    expect(wrapper.find('HeroesHero').at(1).hasClass('selected')).toEqual(true);
  });
});
