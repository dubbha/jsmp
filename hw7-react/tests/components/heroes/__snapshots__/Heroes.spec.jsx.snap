// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heroes should render heroes list 1`] = `
<div
  className="heroes"
>
  <h2>
    My Heroes
  </h2>
  <ul
    className="heroes"
  >
    <HeroesHero
      className=""
      hero={
        Object {
          "id": 3,
          "name": "Optimus Drie",
        }
      }
      setSelectedHero={[Function]}
    />
    <HeroesHero
      className=""
      hero={
        Object {
          "id": 4,
          "name": "Optimus Fier",
        }
      }
      setSelectedHero={[Function]}
    />
    <HeroesHero
      className=""
      hero={
        Object {
          "id": 5,
          "name": "Optimus Funf",
        }
      }
      setSelectedHero={[Function]}
    />
  </ul>
</div>
`;
